internal/tunnel/client/raw_session.go

```diff
@@ -46,11 +47,12 @@ type SessionHandler interface {
 // When RawSession.Accept() returns an error, that means the session is dead.
 // Client sessions run over a muxado session.
 type rawSession struct {
-	mux              *muxado.Heartbeat // the muxado session we're multiplexing streams over
-	id               string            // session id for logging purposes
-	handler          SessionHandler    // callbacks to allow the application to handle requests from the server
-	latency          chan time.Duration
-	closeLatencyOnce sync.Once
+	mux     *muxado.Heartbeat // the muxado session we're multiplexing streams over
+	id      string            // session id for logging purposes
+	handler SessionHandler    // callbacks to allow the application to handle requests from the server
+	latency chan time.Duration
+	closed  atomic.Bool
+	cbGroup sync.WaitGroup
```

Comment:
I feel like this juggling might be better modeled with the internal `reliable.Shutdown`, which we could pull out here.

Basically, I think if we do:

```go
func (s *rawSession) Close() {
  if s.shutdown.Shutdown() {
    close(s.latency)
  }
}

func (s *rawSession) onHeartbeat() {
  if timeout {
    s.shutdown.Shutdown()
    return
  }
  s.shutdown.Do(func() {
    // latency
  })
}
```

I think that ends up being identical to the code you have now, but with a slightly cleaner abstraction maybe?

Untested of course, so I'm not 100% confident. I do think having a unit test that reproduces the panic, and thus lets us be sure it's fixed, would be really helpful!

Reply:
I realized that a `RWLock` is a more direct way to represent this roundabout mutual exclusivity. Should be a lot cleaner now!